beluga/include/beluga/algorithm/amcl_core.hpp

```diff
@@ -0,0 +1,240 @@
+  void update_map(map_type map) { sensor_model_.update_map(std::move(map)); }
+
+  /// Update particles based on motion and sensor information.
+  /**
+   * This method performs a particle filter update step using motion and sensor data. It evaluates whether
+   * an update is necessary based on the configured update policy and the force_update flag. If an update
+   * is required, the motion model and sensor model updates are applied to the particles, and the particle
+   * weights are adjusted accordingly. Also, according to the configured resampling policy, the particles
+   * are resampled to maintain diversity and prevent degeneracy.
+   *
+   * \param state_at_measurement State at the measurement time.
+   * \param measurement Measurement data.
+   * \return An optional pair containing the estimated pose and covariance after the update,
+   *         or std::nullopt if no update was performed.
+   */
+  auto update(state_type state_at_measurement, measurement_type measurement) -> std::optional<estimation_type> {
+    if (particles_.empty()) {
+      return std::nullopt;
+    }
+
+    if (!update_policy_(state_at_measurement) && !force_update_) {
+      return std::nullopt;
+    }
+
+    particles_ |= beluga::actions::propagate(
+                      execution_policy_, motion_model_(control_action_window_ << state_at_measurement)) |  //
+                  beluga::actions::reweight(execution_policy_, sensor_model_(std::move(measurement))) |    //
+                  beluga::actions::normalize(execution_policy_);
+
+    const double random_state_probability = random_probability_estimator_(particles_);
+
+    if (resample_policy_(particles_)) {
+      auto random_state = ranges::compose(beluga::make_from_state<particle_type>, get_random_state_fn());
+
+      if (random_state_probability > 0.0) {
+        random_probability_estimator_.reset();
+      }
+
+      particles_ |= beluga::views::sample |
+                    beluga::views::random_intersperse(std::move(random_state), random_state_probability) |
+                    beluga::views::take_while_kld(
+                        spatial_hasher_,        //
+                        params_.min_particles,  //
+                        params_.max_particles,  //
+                        params_.kld_epsilon,    //
+                        params_.kld_z) |
+                    beluga::actions::assign;
+    }
+
+    force_update_ = false;
+    return beluga::estimate(beluga::views::states(particles_), beluga::views::weights(particles_));
+  }
+
+  /// Force a manual update of the particles on the next iteration of the filter.
+  void force_update() { force_update_ = true; }
+
+  /// Gets a callable that will produce a random state.
+  decltype(auto) get_random_state_fn() const {
+    if constexpr (std::is_invocable_v<random_state_generator_type>) {
+      return random_state_generator_;
+    } else {
+      return random_state_generator_(particles_);
+    }
+  }
```

Comment:
@serraramiro1 nit: consider making this private. Also, `get_random_state_generator()` seems more adequate.